sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+        if (!tracer.isEnabled()) {
+            return operation.invoke(completeRequest, requestOptions);
+        }
+        final StreamingChatCompletionsState state
+            = new StreamingChatCompletionsState(captureContent, request, operation, completeRequest, requestOptions);
+
+        final Mono<StreamingChatCompletionsState> resourceSupplier = Mono.fromSupplier(() -> {
+            final StreamingChatCompletionsState resource = state;
+
+            final Context span
+                = tracer.start(rootSpanName(resource.request), new StartSpanOptions(SpanKind.CLIENT), Context.NONE);
+            traceCompletionRequestAttributes(resource.request, span);
+            traceCompletionRequestEvents(resource.request.getMessages(), span);
+            return resource.setSpan(span);
+        });
+
+        final Function<StreamingChatCompletionsState, Flux<StreamingChatCompletionsUpdate>> resourceClosure
+            = resource -> {
+                final RequestOptions rOptions = resource.requestOptions.setContext(resource.span);
+                final Flux<StreamingChatCompletionsUpdate> completionChunks
+                    = resource.operation.invoke(resource.completeRequest, rOptions);
+                return completionChunks.doOnNext(resource::onNextChunk);
+            };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncComplete = resource -> {
+            final Context span = resource.span;
+            final StreamingChatCompletionsUpdate lastChunk = resource.lastChunk;
+            final String finishReasons = resource.getFinishReasons();
+
+            traceCompletionResponseAttributes(lastChunk, finishReasons, span);
+            traceChoiceEvent(resource.toJson(logger), OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, null, span);
+            return Mono.empty();
+        };
+
+        final BiFunction<StreamingChatCompletionsState, Throwable, Mono<Void>> asyncError = (resource, throwable) -> {
+            final Context span = resource.span;
+
+            tracer.setAttribute("error.type", throwable.getClass().getName(), span);
+            traceChoiceEvent(FINISH_REASON_ERROR, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, throwable, span);
+            return Mono.empty();
+        };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncCancel = resource -> {
+            final Context span = resource.span;
+            tracer.setAttribute("error.type", "cancelled", span);
+            traceChoiceEvent(FINISH_REASON_CANCELED, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end("cancelled", null, span);
+            return Mono.empty();
+        };
+
+        return Flux.usingWhen(resourceSupplier, resourceClosure, asyncComplete, asyncError, asyncCancel);
+    }
+
+    //<editor-fold desc="Private util types, methods">
+    private String rootSpanName(ChatCompletionsOptions completeRequest) {
```

Comment:
nit: it's not usually a root span, it's just a span - if chat call happens in the scope of incoming request (or SB processor callback), it'd be a child of that span.

Reply:
I see, I renamed it to `spanName(..)`